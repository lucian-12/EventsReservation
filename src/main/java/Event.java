public class Event {

    private String name;
    private String type;

    public Event(String name, String type) {
        this.name = name;
        this.type = type;
    }

    public String getName() {
        return name;
    }

    public Event setName(String name) {
        this.name = name;
        return this;
    }

    public String getType() {
        return type;
    }

    public Event setType(String type) {
        this.type = type;
        return this;
    }
}
