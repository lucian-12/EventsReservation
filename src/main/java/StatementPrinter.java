import java.text.NumberFormat;
import java.util.Locale;

public class StatementPrinter {

    public String print(Invoice invoice) {
        int totalAmount = 0;
        int volumeCredits = 0;
        StringBuilder result = new StringBuilder(String.format("Statement for %s\n", invoice.getCustomer()));

        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);

        for (Reservation reservation : invoice.getReservations()) {
            int thisAmount = amountFor(reservation);

            volumeCredits += volumeCreditsFor(reservation);

            // print line for this order
            result.append(String.format("  %s: %s (%s seats)\n", reservation.getEvent().getName(), frmt.format(thisAmount / 100), reservation.getNbSeats()));
            totalAmount += thisAmount;
        }
        result.append(String.format("Amount owed is %s\n", frmt.format(totalAmount / 100)));
        result.append(String.format("You earned %s credits\n", volumeCredits));
        return result.toString();
    }

    private int volumeCreditsFor(Reservation reservation) {
        int result = 0;
        // add volume credits
        result += Math.max(reservation.getNbSeats() - 2, 0);
        // add extra credit for every workshop attendee
        if ("workshop".equals(reservation.getEvent().getType()))
            result += Math.floor(reservation.getNbSeats());
        return result;
    }

    private int amountFor(Reservation reservation) {
        int result;

        switch (reservation.getEvent().getType()) {
            case "conference":
                result = 40000;
                if (reservation.getNbSeats() > 3) {
                    result += 1000 * (reservation.getNbSeats() - 3);
                }
                break;
            case "workshop":
                result = 30000;
                if (reservation.getNbSeats() > 2) {
                    result += 10000 + 500 * (reservation.getNbSeats() - 2);
                }
                result += 300 * reservation.getNbSeats();
                break;
            default:
                throw new Error("unknown type: ${event.type}");
        }
        return result;
    }

}
