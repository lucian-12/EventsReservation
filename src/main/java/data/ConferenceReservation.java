package data;

public class ConferenceReservation extends Reservation {
    public ConferenceReservation(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 40000;
        if (getNbSeats() > 3) {
            result += 1000 * (getNbSeats() - 3);
        }
        return result;
    }

    @Override
    int calculateVolumeCredits() {
        int result = 0;
        result += Math.max(getNbSeats() - 2, 0);
        return result;
    }
}
