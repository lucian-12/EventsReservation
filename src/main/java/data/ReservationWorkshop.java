package data;

public class ReservationWorkshop extends Reservation {
    public ReservationWorkshop(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 30000;
        if (getNbSeats() > 2) {
            result += 10000 + 500 * (getNbSeats() - 2);
        }
        result += 300 * getNbSeats();
        return result;
    }

    @Override
    int calculateVolumeCredits() {
        int result = 0;
        result += Math.max(getNbSeats() - 2, 0);
        result += Math.floor(getNbSeats());
        return result;
    }
}
