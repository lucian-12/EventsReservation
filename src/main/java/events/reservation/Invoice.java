package events.reservation;

import java.util.List;

public class Invoice {

    private String customer;
    private List<Reservation> reservations;

    public Invoice(String customer, List<Reservation> reservations) {
        this.customer = customer;
        this.reservations = reservations;
    }

    public String getCustomer() {
        return customer;
    }

    public Invoice setCustomer(String customer) {
        this.customer = customer;
        return this;
    }

    public List<Reservation> getReservations() {
        return reservations;
    }

    public Invoice setReservations(List<Reservation> reservations) {
        this.reservations = reservations;
        return this;
    }
}
