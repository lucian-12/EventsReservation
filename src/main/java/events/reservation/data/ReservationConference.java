package events.reservation.data;

public class ReservationConference extends Reservation {
    public ReservationConference(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 40000;
        if (getNbSeats() > 3) {
            result += 1000 * (getNbSeats() - 3);
        }
        return result;
    }

    @Override
    public int calculateTotalVolumeCredits() {
        return Math.max(getNbSeats() - 2, 0);
    }
}
