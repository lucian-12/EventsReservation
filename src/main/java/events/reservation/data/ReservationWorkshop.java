package events.reservation.data;

public class ReservationWorkshop extends Reservation {
    public ReservationWorkshop(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 30000;
        if (getNbSeats() > 2) {
            result += 10000 + 500 * (getNbSeats() - 2);
        }
        result += 300 * getNbSeats();
        return result;
    }

}
