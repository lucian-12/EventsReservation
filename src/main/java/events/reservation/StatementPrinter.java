package events.reservation;

import events.reservation.data.StatementData;

import java.text.NumberFormat;
import java.util.Locale;

public class StatementPrinter {

    public String print(Invoice invoice) {
        StatementData statementData = invoice.createStatementData();
        return renderPlainText(statementData);
    }

    private String renderPlainText(StatementData statementData) {
        StringBuilder result = new StringBuilder(String.format("Statement for %s\n",
                statementData.getCustomer()));

        for (Reservation reservation : statementData.getReservations()) {

            // print line for this order
            result.append(String.format("  %s: %s (%s seats)\n",
                    reservation.getEvent().getName(),
                    usd(reservation.calculateAmount()),
                    reservation.getNbSeats()));
        }

        result.append(String.format("Amount owed is %s\n", usd(statementData.getCalculateTotalAmount())));
        result.append(String.format("You earned %s credits\n", statementData.getCalculateTotalVolumeCredits()));
        return result.toString();
    }

    private String usd(int i) {
        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);
        return frmt.format(i / 100);
    }

}
