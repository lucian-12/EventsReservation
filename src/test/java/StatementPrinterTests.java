import data.ReservationConference;
import data.Invoice;
import data.ReservationWorkshop;
import data.Event;
import org.junit.jupiter.api.Test;
import view.StatementPrinter;

import java.util.List;

import static org.approvaltests.Approvals.verify;

public class StatementPrinterTests {

    @Test
    void exampleStatement() {

        Invoice invoice = new Invoice("BigCo", List.of(
                new ReservationConference("dev", 5, new Event("DevConference", "conference")),
                new ReservationWorkshop("web", 3, new Event("WebWorkshop", "workshop")),
                new ReservationConference("ml", 4, new Event("MachineLearning", "conference"))));

        StatementPrinter statementPrinter = new StatementPrinter();
        String result = statementPrinter.print(invoice);

        verify(result);
    }
}
