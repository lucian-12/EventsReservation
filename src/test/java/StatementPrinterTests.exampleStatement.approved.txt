Statement for BigCo
  DevConference: $420.00 (5 seats)
  WebWorkshop: $414.00 (3 seats)
  DevOps: $410.00 (4 seats)
Amount owed is $1,244.00
You earned 9 credits